### ISSUE TITLE
Accessibility concerns with non-fallback canvas children

### ISSUE THREAD

**Issue by schenney-chromium** [WICG/html-in-canvas#11]
Elements in fallback content which are passed to `placeElement()` are still exposed to assistive technology APIs, just like all other canvas fallback content. There needs to be a way to indicate that some children are not yet in the canvas (until `placeElement()` is called on them) and hence should not be visible to assistive technologies.

This can be achieved with any of the usual techniques for hiding content from assistive technology, such as using the hidden, inert or aria-hidden attributes, or applying a display: none or display: hidden style. Browsers cannot automate this because they do not know which fallback content is for non-placeElement portions of the canvas, and hence should remain accessible.

Some mechanism could be defined for distinguishing the two different use cases for fallback content and should at least be noted in the spec. We may wish to define a specific method of hiding so that the attribute/property can be removed when `placeElement()` is called.

**Comment by alice** [comment]
Thoughts from the session just now:

@chrishtr suggested having modes where fallback content which isn't placed is/isn't exposed to AT. 

That might be a start; perhaps you could have `placeElement()` etc. cause the fallback content to not be exposed by default (with content shown on the canvas always exposed), and you can opt in (maybe an attribute on the `<canvas>` or something) to saying "yes, I intend this fallback content to be exposed as well".

Another alternative might be, to potentially solve a more narrow problem, if there are a set of elements which are intended to replace one another with some being considered "stale", to offer authors a way to declare those elements as mutually exclusive on the canvas, so that placing one will remove the others, similar to how re-placing an element will cause the earlier drawing to be removed. 

That may be completely infeasible; I just wanted to continue thinking through the problem a little more.

**Comment by schenney-chromium** [comment]
Alice and I chatted again today.

The current proposal for handling hit testing is to mirror the canvas Current Transform Matrix (CTM) plus the drawElement position into the CSS transform style of the drawn child. This can be thought of as moving the DOM content to an invisible layer in front of the canvas, such that it captures the events before they get to the canvas. Focus etc would all be in the right position in the canvas. `drawElement` in this model is like making the DOM content visible in the canvas.

We agree that this is an improvement over the existing canvas fallback content for accessibility primarily because it gives developers additional clues as to the state of the fallback content: it will appear in DevTools and it will respond to events. Developers use DevTools, so they are reasonably likely to see the content there. And messed up event handling is sure to be noticed.

Another benefit is improved ease of use for focus APIs that already exist in canvas. You could just define the focus path as the current client rect of the DOM.

It would be preferable to automatically set the transform style from the canvas transform at the time it is drawn. Doing so would make it harder to mess things up.

We will need some way to avoid breaking existing applications, so we should consider an explicit attribute to mark canvas children as participating in event handling. If we rename the `layout-subtree` attribute to `capture-events` and put it on children instead of the canvas we can leave existing sites exactly as is.

Alice pointed out a potentially better solution for Canvas a11y to address some of the challenges people currently have building good a11y. Best practice now for "entire app as canvas" like Google docs is to build an offscreen DOM structure just for a11y, and make use of aria labels to guide assistive technologies to the right place. Maybe we could have an explicit `placeAccessibility` API call to associate ARIA content to a canvas area without drawing it or making it respond to pointer events.

No matter the outcome, we should consider how these existing canvas a11y solutions might be more easily done in the `drawElement` framework.

**Comment by progers** [comment]
Thanks for thinking about this!

> We will need some way to avoid breaking existing applications

As long as html-in-canvas has an opt-in of some form (currently the `layoutsubtree` attribute), no content will break by default, but I agree we need to think about the upgrade paths of existing applications.

> Maybe we could have an explicit placeAccessibility API call to associate ARIA content to a canvas area without drawing it or making it respond to pointer events.

We have some imperfect data about existing fallback content with the following:
[HTMLCanvasElement](https://chromestatus.com/metrics/feature/timeline/popularity/1503) - measures pages with canvas - 45%
[CanvasFallbackContent](https://chromestatus.com/metrics/feature/timeline/popularity/5608) - measures pages with text nodes OR element nodes elements below canvas - 0.4%
[CanvasFallbackElementContent](https://chromestatus.com/metrics/feature/timeline/popularity/5609) - measures pages with element nodes below canvas - 0.15%

I looked at 10 pages at random from the HTTPArchive with fallback element content, and 8 used the form `<canvas><p>fallback text</p></canvas>`, and two had actual elements under canvas (https://www.dslweb.de and https://www.gardatrentino.it/en).

So, though it is rare, fallback dom seems like another example where we would want to have content under canvas that is visible to accessibility but not automatically hit testable. This is in addition to the case mentioned above, of moving existing offscreen a11y dom content under canvas.


This issue of making content visible only to accessibility but not hit testing seems like a large issue. Sites are currently doing this with `.sr-only` styles (Miro is doing this exactly, and Google Docs is effectively doing it), and there have been proposals for standardizing this (e.g., [visually-hidden](https://github.com/w3c/csswg-drafts/issues/560)), but there are issues (misuse, disjoint usecases, etc). We could have an explicit API for disabling the automatic hit testing of certain canvas children (e.g., using `placeAccessibility(...)`, or _not_ using `capture-events`), but I wonder if it would be better to not have an html-in-canvas special-case for this, for the same issues (misuse, etc)? Said another way, html-in-canvas enables customizing the way an element is rendered while preserving as much of the standard browser behavior as possible (accessibility, hit testing, translate, forms, etc).

**Comment by alice** [comment]
I just want to quickly note that "hit testing" is ambiguous - there is accessibility-specific hit testing which is used for touch exploration (and, I learned only yesterday, mouse exploration), which is why placing elements for accessibility would be useful, and we should make sure that we're not ruling out accessibility-specific hit testing by ruling elements out for DOM hit testing. (This has been a problem with `pointer-events: none`, I hear.)

**Comment by chrishtr** [comment]
The current API's semantics are:
* When `layoutsubtree` is not present on the `<canvas>`, a11y works exactly as it does today on the existing web
* When  `layoutsubtree` *is* present, all descendants will be hit tested and present in a11y. If a developer desires to adjust that, they should use features like `pointer-events` and `aria-hidden`.
